### PR TITLE
Implement repo uberclone.

### PR DESCRIPTION
**Pull request by crummel**
This adds a new darc verb, `clone`, that will either clone a specified remote repo and all repos in its dependency graph, or look at a local Version.Details.xml and clone all repos in it and all repos in their dependency graph.  It will put all the repos in the same directory with the format `{name}.{sha}`.

### CODE REVIEW

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs [hunk 1]
```diff
@@ -0,0 +1,204 @@
+using CSharpx;
```
Comment (mmitche) [inline src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs]: Nit: Missing license header.

File: src/Microsoft.DotNet.Darc/src/Darc/Options/CloneCommandLineOptions.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+using CommandLine;
+using Microsoft.DotNet.Darc.Operations;
+
+namespace Microsoft.DotNet.Darc.Options
+{
+    [Verb("clone", HelpText = "clone a remote repo and all of its dependency repos")]
+
+    internal class CloneCommandLineOptions : CommandLineOptions
+    {
+        [Option('l', "local", HelpText = "Clone all repos that this local repository declares a reference on.")]
```
Comment (mmitche) [inline src/Microsoft.DotNet.Darc/src/Darc/Options/CloneCommandLineOptions.cs]: Local in the context of the other commands typically means "don't touch external resources".  I would actually just remove this option and imply it if repo and version is not passed in.

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+            _options = options;
+        }
+
+        public override async Task<int> ExecuteAsync()
+        {
+            try
+            {
+                EnsureOptionsCompatibility(_options);
+                RemoteFactory remoteFactory = new RemoteFactory(_options);
+                HashSet<StrippedDependency> seenDependencies = new HashSet<StrippedDependency>();
```
Comment (mmitche) [inline src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs]: I don't think you actually need this.  Your cloning code skips a directory if it already exists, so shouldn't it do the "already been cloned" filtering by default?

Comment (crummel) [reply src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs]: Yeah, this was leftover from look at dependencies in repos after cloning them.  Removed.

File: src/Microsoft.DotNet.Darc/src/Darc/Options/CloneCommandLineOptions.cs (hunk 2 above)
Comment (crummel) [reply src/Microsoft.DotNet.Darc/src/Darc/Options/CloneCommandLineOptions.cs]: Removed as suggested.

File: src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs (hunk 1 above)
Comment (crummel) [reply src/Microsoft.DotNet.Darc/src/Darc/Operations/CloneOperation.cs]: Added.

**Comment by jcagme** [comment]
I'll take a look today but not be blocked by my review. If there is something that requires a change I'll let you know

**Pull request merged**